vlanBridge.go

```diff
@@ -287,3 +314,118 @@ func (vl *VlanBridge) initFgraph() error {
+ * Src EP not known, Dest EP known:
+ *      - Proxy ARP if Dest EP is present locally on the host
+ * Src and Dest EP not known: 
+ *      - Ignore processing the request
+ */
+func (vl *VlanBridge) processArp(pkt protocol.Ethernet, inPort uint32) {
+	switch t := pkt.Data.(type) {
+	case *protocol.ARP:
+		log.Debugf("Processing ARP packet on port %d: %+v", inPort, *t)
+		var arpIn protocol.ARP = *t
+
+		switch arpIn.Operation {
+		case protocol.Type_Request:
+			// Lookup the Source and Dest IP in the endpoint table
+			srcEp := vl.agent.getEndpointByIp(arpIn.IPSrc)
+			dstEp := vl.agent.getEndpointByIp(arpIn.IPDst)
+
+			// No information about the src or dest EP. Ignore processing.
+			if srcEp == nil && dstEp == nil {
+				log.Debugf("No information on source/destination. Ignoring ARP request.")
+				return
+			}
+			// If we know the dstEp to be present locally, send the Proxy ARP response
+			if dstEp != nil {
+				// Container to Container communication. Send proxy ARP response.
+				// Unknown node to Container communication
+				//   -> Send proxy ARP response only if Endpoint is local.
+				//   -> This is to avoid sending ARP responses from ofnet agent on multiple hosts
+				if srcEp != nil ||
+					(srcEp == nil && dstEp.OriginatorIp.String() == vl.agent.localIp.String()) {
+					// Form an ARP response
+					arpPkt, _ := protocol.NewARP(protocol.Type_Reply)
+					arpPkt.HWSrc, _ = net.ParseMAC(dstEp.MacAddrStr)
+					arpPkt.IPSrc = arpIn.IPDst
+					arpPkt.HWDst = arpIn.HWSrc
+					arpPkt.IPDst = arpIn.IPSrc
+					log.Debugf("Sending Proxy ARP response: %+v", arpPkt)
+
+					// Build the ethernet packet
+					ethPkt := protocol.NewEthernet()
+					ethPkt.VLANID.VID = pkt.VLANID.VID
+					ethPkt.HWDst = arpPkt.HWDst
+					ethPkt.HWSrc = arpPkt.HWSrc
+					ethPkt.Ethertype = 0x0806
+					ethPkt.Data = arpPkt
+					log.Debugf("Sending Proxy ARP response Ethernet: %+v", ethPkt)
+
+					// Construct Packet out
+					pktOut := openflow13.NewPacketOut()
+					pktOut.Data = ethPkt
+					pktOut.AddAction(openflow13.NewActionOutput(inPort))
+
+					// Send the packet out
+					vl.ofSwitch.Send(pktOut)
+
+					return
+				}
+			}
+			if srcEp != nil && dstEp == nil {
+				// ARP request from local container to unknown IP
+				// Reinject ARP to uplinks
+
+				ethPkt := protocol.NewEthernet()
+				ethPkt.VLANID.VID = srcEp.Vlan
+				ethPkt.HWDst = pkt.HWDst
+				ethPkt.HWSrc = pkt.HWSrc
+				ethPkt.Ethertype = 0x0806
+				ethPkt.Data = &arpIn
+
+				log.Infof("Received ARP request for unknown IP: %v."+
+					"Reinjecting ARP request Ethernet to uplinks: %+v", arpIn.IPDst, ethPkt)
+
+				// Packet out
+				pktOut := openflow13.NewPacketOut()
+				pktOut.InPort = inPort
+				pktOut.Data = ethPkt
+				for _, portNo := range vl.uplinkDb {
+					log.Debugf("Sending to uplink: %+v", portNo)
+					pktOut.AddAction(openflow13.NewActionOutput(portNo))
+				}
+
+				// Send the packet out
+				vl.ofSwitch.Send(pktOut)
+			}
+
+		case protocol.Type_Reply:
```

Comment:
@abhinandanpb 
The ARP reply packets do not come to the controller. The processing here though is just for future reference/processing.